core/src/main/java/org/apache/calcite/util/mapping/Mappings.java

```diff
@@ -1245,12 +1271,30 @@ public void set(int source, int target) {
       assert isValid();
     }
 
+    /**
+     * Returns the source that a target maps to, or -1 if it is not mapped.
+     *
+     *  @throws NoElementException if target is out of sources range.
+     */
     public int getSourceOpt(int target) {
-      return sources[target];
+      try {
+        return sources[target];
+      } catch (ArrayIndexOutOfBoundsException e) {
+        throw new Mappings.NoElementException("invalid target " + target);
```

Comment:
According to the doc, NoElementException should be when a mapping is expected to return one element but returns none. -1 means it is not mapped. Maybe we can just leave it as it is (and other places), let it throw ArrayIndexOutOfBoundsException.

Reply:
Hmm.. it is also said in javadoc that `FunctionMapping#getTarget` throws `NoElementException if source is not mapped`. Which is closer to the truth. In this case `NoElementException` is valid move.

I also noticed that for some mappings it was an original behavior, i.e. `Permutation#getTarget()`.  I just made the fix by analogy with this legacy behavior.
I agree that it is far from perfect. May be `IllegalArgumentException` is more suitable here. But I think that `ArrayIndexOutOfBoundsException` is not the best choice.

What if I just change the javadoc of `NoElementException` to `Should be thrown when source or target is not mapped`?

Reply:
Checking `AbstractMapping`, NoElementException is thrown when `getTargetOpt` or `getSourceOpt` returns -1, which is aligning with the javadoc.

Reply:
So, what do you propose? Just throw `ArrayIndexOutOfBoundsException`?

Reply:
yes.

Reply:
I don't mind throwing `NoElementException`, if the desc in the javadoc is misleading. @julianhyde Can you confirm?

Reply:
`NoElementException` is for the case where the source is within the domain (0 .. sourceCount - 1) but no mapping is present.

If the source is outside the domain - less than 0 or greater than or equal to sourceCount - then `IndexOutOfBoundsException` would be appropriate.

I don't think `getSourceOpt` or `getTargetOpt` should ever throw `NoElementException`. They have a different way to communicate no element - namely by returning -1.

Reply:
I changed exception to `IndexOutOfBoundsException` when source/target is out of domain. Please, take a look.